File: retaguarda-software-house/java/retaguarda_sh/src/main/java/com/t2ti/retaguarda_sh/controller/acbr/AcbrMonitorController.java
/*******************************************************************************
Title: T2Ti ERP 3.0                                                                
Description: Controller relacionado à tabela [NFE_CONFIGURACAO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
package com.t2ti.retaguarda_sh.controller.acbr;

import java.io.File;

import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.FileSystemResource;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.t2ti.retaguarda_sh.exception.AcbrMonitorException;
import com.t2ti.retaguarda_sh.exception.ExcecaoGenericaServidorException;
import com.t2ti.retaguarda_sh.model.transiente.ObjetoNfe;
import com.t2ti.retaguarda_sh.service.acbr.AcbrMonitorService;
import com.t2ti.retaguarda_sh.util.Biblioteca;

@RestController
@RequestMapping(value = "/acbr-monitor", produces = "application/json;charset=UTF-8")
public class AcbrMonitorController {

	@Autowired
	private AcbrMonitorService service;
	
	@PostMapping(value = "/emite-nfce", produces = "application/pdf")
	public FileSystemResource emitirNfce(@RequestBody String corpoRequisicao, @RequestHeader("numero") String numero, @RequestHeader("cnpj") String cnpj, HttpServletResponse response) {
		boolean erroMonitor = false;
		String retorno = "";
		try {
			String nfceIni = Biblioteca.decifrar(corpoRequisicao);
			retorno = service.emitirNfce(Biblioteca.decifrar(numero), Biblioteca.decifrar(cnpj), nfceIni);
			if (!retorno.contains("ERRO")) {
			    return new FileSystemResource(new File(retorno));							
			} else {
				erroMonitor = true;
				throw new Exception();
			}
		} catch (Exception e) {
			if (erroMonitor) {
				throw new AcbrMonitorException(retorno);
			} else {
				throw new ExcecaoGenericaServidorException(
						"Erro no Servidor [Emitir NFC-e] - Exceção: " + e.getMessage());				
			}
		}
	}
	
	@PostMapping(value = "/emite-nfce-contingencia", produces = "application/pdf")
	public FileSystemResource emitirNfceContingencia(@RequestBody String corpoRequisicao, @RequestHeader("numero") String numero, @RequestHeader("cnpj") String cnpj, HttpServletResponse response) {
		boolean erroMonitor = false;
		String retorno = "";
		try {
			String nfceIni = Biblioteca.decifrar(corpoRequisicao);
			retorno = service.emitirNfceContingencia(Biblioteca.decifrar(numero), Biblioteca.decifrar(cnpj), nfceIni);
			if (!retorno.contains("ERRO")) {
			    return new FileSystemResource(new File(retorno));							
			} else {
				erroMonitor = true;
				throw new Exception();
			}
		} catch (Exception e) {
			if (erroMonitor) {
				throw new AcbrMonitorException(retorno);
			} else {
				throw new ExcecaoGenericaServidorException(
						"Erro no Servidor [Emitir NFC-e Contingência] - Exceção: " + e.getMessage());
			}
		}
	}
	
	@PostMapping(value = "/transmite-nfce-contingenciada", produces = "application/pdf")
	public FileSystemResource transmitirNfceContingenciada(@RequestHeader("chave") String chave, @RequestHeader("cnpj") String cnpj, HttpServletResponse response) {
		boolean erroMonitor = false;
		String retorno = "";
		try {
			retorno = service.transmitirNfceContingenciada(Biblioteca.decifrar(chave), Biblioteca.decifrar(cnpj));
			if (!retorno.contains("ERRO")) {
			    return new FileSystemResource(new File(retorno));							
			} else {
				erroMonitor = true;
				throw new Exception();
			}
		} catch (Exception e) {
			if (erroMonitor) {
				throw new AcbrMonitorException(retorno);
			} else {
				throw new ExcecaoGenericaServidorException(
						"Erro no Servidor [Transmitir NFC-e Contingenciada] - Exceção: " + e.getMessage());
			}
		}
	}
	
	@PostMapping(value = "/trata-nota-anterior-contingencia")
	public String tratarNotaAnteriorContingencia(@RequestBody String corpoRequisicao) {
		ObjectMapper objectMapper = new ObjectMapper();
		try {
	        ObjetoNfe objetoNfe = objectMapper.readValue(Biblioteca.decifrar(corpoRequisicao), ObjetoNfe.class);
	        String retorno = service.tratarNotaAnteriorContingencia(objetoNfe);
	        return Biblioteca.cifrar(retorno);
		} catch (Exception e) {
			throw new ExcecaoGenericaServidorException("Erro no Servidor [Tratar Nota Anterior Contingencia] - Exceção: " + e.getMessage());
		}
	}
	
	@PostMapping(value = "/inutiliza-numero-nota")
	public String inutilizarNumeroNota(@RequestBody String corpoRequisicao) {
		ObjectMapper objectMapper = new ObjectMapper();
		try {
	        ObjetoNfe objetoNfe = objectMapper.readValue(Biblioteca.decifrar(corpoRequisicao), ObjetoNfe.class);
	        String retorno = service.inutilizarNumero(objetoNfe);
	        return Biblioteca.cifrar(retorno);
		} catch (Exception e) {
			throw new ExcecaoGenericaServidorException("Erro no Servidor [Inutilizar Numero Nota] - Exceção: " + e.getMessage());
		}
	}

	@PostMapping(value = "/cancela-nfce")
	public String cancelarNfce(@RequestBody String corpoRequisicao) {
		ObjectMapper objectMapper = new ObjectMapper();
		try {
	        ObjetoNfe objetoNfe = objectMapper.readValue(Biblioteca.decifrar(corpoRequisicao), ObjetoNfe.class);
	        String retorno = service.cancelarNfce(objetoNfe);
	        return Biblioteca.cifrar(retorno);
		} catch (Exception e) {
			throw new ExcecaoGenericaServidorException("Erro no Servidor [Cancelar NFC-e] - Exceção: " + e.getMessage());
		}
	}
	
	@PostMapping(value = "/gera-pdf-danfe-nfce", produces = "application/pdf")
	public FileSystemResource gerarPdfDanfeNfce(@RequestHeader("chave") String chave, @RequestHeader("cnpj") String cnpj, HttpServletResponse response) {
		boolean erroMonitor = false;
		String retorno = "";
		try {
			retorno = service.gerarPdfDanfeNfce(Biblioteca.decifrar(chave), Biblioteca.decifrar(cnpj));
			if (!retorno.contains("ERRO")) {
			    return new FileSystemResource(new File(retorno));							
			} else {
				erroMonitor = true;
				throw new Exception();
			}
		} catch (Exception e) {
			if (erroMonitor) {
				throw new AcbrMonitorException(retorno);
			} else {
				throw new ExcecaoGenericaServidorException(
						"Erro no Servidor [Gerar Pdf Danfe Nfce] - Exceção: " + e.getMessage());
			}
		}
	}
		
	@GetMapping(value = "/download-xml-periodo", produces = "application/zip")
	public FileSystemResource retornarArquivosXmlPeriodo(@RequestHeader("periodo") String periodo, @RequestHeader("cnpj") String cnpj) {
		try {
			boolean retorno = service.gerarZipArquivosXml(Biblioteca.decifrar(periodo), Biblioteca.decifrar(cnpj));
			if (retorno) {
			    String caminhoArquivo = "C:\\ACBrMonitor\\" + cnpj + "\\NotasFiscaisNFCe_" + periodo + ".zip";
			    return new FileSystemResource(new File(caminhoArquivo));							
			} else {				
				throw new Exception();
			}
		} catch (Exception e) {
			throw new ExcecaoGenericaServidorException(
					"Erro no Servidor [Retornar XML Período] - Exceção: " + e.getMessage());
		}
	}

	@PostMapping(value = "/atualiza-certificado")
	public String atualizarCertificado(@RequestBody String corpoRequisicao, @RequestHeader("hash-registro") String senha, @RequestHeader("cnpj") String cnpj) {
		try {
			String certificadoBase64 = Biblioteca.decifrar(corpoRequisicao);
			service.atualizarCertificado(certificadoBase64, Biblioteca.decifrar(senha), Biblioteca.decifrar(cnpj));			
			return "Certificado atualizado com sucesso.";
		} catch (Exception e) {
			throw new ExcecaoGenericaServidorException("Erro no Servidor [Atualizar Certificado] - Exceção: " + e.getMessage());
		}
	}

	
}

File: retaguarda-software-house/php/retaguarda-sh/src/controller/acbr/AcbrMonitorController.php
<?php
/*******************************************************************************
Title: T2Ti ERP 3.0
Description: Controller relacionado ao ACBrMonitor
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
use Slim\Psr7\Stream;

class AcbrMonitorController extends ControllerBase
{

    public function emitirNfce($request, $response, $args)
    {
        try {
            $nfceIni = Biblioteca::decifrar($request->getBody());
            $cnpj = Biblioteca::decifrar($request->getHeaders()['cnpj'][0]);
            $numero = Biblioteca::decifrar($request->getHeaders()['numero'][0]);
                        
			$retorno = AcbrMonitorService::emitirNfce($numero, $cnpj, $nfceIni);

            if (!str_contains($retorno, "ERRO")) {
                $fh = fopen($retorno, 'rb');
                $stream = new Stream($fh); 
                return parent::retornarArquivo($response, $stream, 'application/pdf');                
            } else {
                return parent::tratarErro($response, 418, $retorno, null);                
            }
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

    public function emitirNfceContingencia($request, $response, $args)
    {
        try {
            $nfceIni = Biblioteca::decifrar($request->getBody());
            $cnpj = Biblioteca::decifrar($request->getHeaders()['cnpj'][0]);
            $numero = Biblioteca::decifrar($request->getHeaders()['numero'][0]);
                        
			$retorno = AcbrMonitorService::emitirNfceContingencia($numero, $cnpj, $nfceIni);

            if (!str_contains($retorno, "ERRO")) {
                $fh = fopen($retorno, 'rb');
                $stream = new Stream($fh); 
                return parent::retornarArquivo($response, $stream, 'application/pdf');                
            } else {
                return parent::tratarErro($response, 418, $retorno, null);                
            }
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

    public function transmitirNfceContingenciada($request, $response, $args)
    {
        try {
            $cnpj = Biblioteca::decifrar($request->getHeaders()['cnpj'][0]);
            $chave = Biblioteca::decifrar($request->getHeaders()['chave'][0]);
                        
			$retorno = AcbrMonitorService::transmitirNfceContingenciada($chave, $cnpj);

            if (!str_contains($retorno, "ERRO")) {
                $fh = fopen($retorno, 'rb');
                $stream = new Stream($fh); 
                return parent::retornarArquivo($response, $stream, 'application/pdf');                
            } else {
                return parent::tratarErro($response, 418, $retorno, null);                
            }
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

    public function tratarNotaAnteriorContingencia($request, $response, $args)
    {
        try {
            $objetoNfe = json_decode(Biblioteca::decifrar($request->getBody()));

            $retorno = AcbrMonitorService::tratarNotaAnteriorContingencia($objetoNfe);

            $response->getBody()->write(Biblioteca::cifrar($retorno));
            return $response
                ->withHeader('Content-Type', 'application/json');
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

    public function inutilizarNumeroNota($request, $response, $args)
    {
        try {
            $objetoNfe = json_decode(Biblioteca::decifrar($request->getBody()));

            $retorno = AcbrMonitorService::inutilizarNumero($objetoNfe);

            $response->getBody()->write(Biblioteca::cifrar($retorno));
            return $response
                ->withHeader('Content-Type', 'application/json');
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

    public function cancelarNfce($request, $response, $args)
    {
        try {
            $objetoNfe = json_decode(Biblioteca::decifrar($request->getBody()));

            $retorno = AcbrMonitorService::cancelarNfce($objetoNfe);

            $response->getBody()->write(Biblioteca::cifrar($retorno));
            return $response
                ->withHeader('Content-Type', 'application/json');
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

    public function gerarPdfDanfeNfce($request, $response, $args)
    {
        try {
            $cnpj = Biblioteca::decifrar($request->getHeaders()['cnpj'][0]);
            $chave = Biblioteca::decifrar($request->getHeaders()['chave'][0]);
                        
			$retorno = AcbrMonitorService::gerarPdfDanfeNfce($chave, $cnpj);

            if (!str_contains($retorno, "ERRO")) {
                $fh = fopen($retorno, 'rb');
                $stream = new Stream($fh); 
                return parent::retornarArquivo($response, $stream, 'application/pdf');                
            } else {
                return parent::tratarErro($response, 418, $retorno, null);                
            }
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

    public function retornarArquivosXmlPeriodo($request, $response, $args)
    {
        try {
            $cnpj = Biblioteca::decifrar($request->getHeaders()['cnpj'][0]);
            $periodo = Biblioteca::decifrar($request->getHeaders()['periodo'][0]);

            $retorno = AcbrMonitorService::gerarZipArquivosXml($periodo, $cnpj);

            if ($retorno) {
                $file = "C:\\ACBrMonitor\\" . $cnpj . "\\NotasFiscaisNFCe_" . $periodo . ".zip";
                $fh = fopen($file, 'rb');
                $stream = new Stream($fh); 
                return parent::retornarArquivo($response, $stream, 'application/zip');                    
            } else {
                return $response
                ->withStatus(500)
                ->withBody('Problemas na criação do arquivo [Retornar XML Período]')
                ->withHeader('Content-Type', 'application/json');
            }

		} catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Retornar XML Período]', $e);
        }
    }

    public function atualizarCertificado($request, $response, $args)
    {
        try {
            // pegaa a string base 64 do corpo
            $certificadoBase64 = Biblioteca::decifrar($request->getBody());

            $cnpj = Biblioteca::decifrar($request->getHeaders()['cnpj'][0]);
            $senha = Biblioteca::decifrar($request->getHeaders()['hash-registro'][0]);
                        
			// chama o método para atualizar o certificado
			AcbrMonitorService::atualizarCertificado($certificadoBase64, $senha, $cnpj);

            $retorno = 'Certificado atualizado com sucesso.';
            $response->getBody()->write($retorno);

            return $response
                ->withStatus(200)
                ->withHeader('Content-Type', 'application/json');
        } catch (Exception $e) {
            return parent::tratarErro($response, 500, 'Erro no Servidor [Atualizar AcbrMonitor]', $e);
        }
    }

}
